File: snippets/python/config-loader.py
"""
USE: When loading and validating configuration from environment variables
REQUIRES: pydantic>=2.0, pydantic-settings>=2.0, Python 3.10+
PATTERN: Configuration, Validation
"""

from enum import Enum
from functools import lru_cache
from typing import Any

from pydantic import Field, field_validator, model_validator
from pydantic_settings import BaseSettings, SettingsConfigDict


class Environment(str, Enum):
    """Application environment."""
    DEVELOPMENT = "development"
    STAGING = "staging"
    PRODUCTION = "production"
    TEST = "test"


class LogLevel(str, Enum):
    """Logging level."""
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"


# CUSTOMIZE: Define your configuration schema


class DatabaseSettings(BaseSettings):
    """Database configuration."""
    model_config = SettingsConfigDict(env_prefix="DB_")

    url: str = Field(
        ...,
        description="Database connection URL",
        examples=["postgresql://user:pass@localhost:5432/dbname"],
    )
    pool_size: int = Field(
        default=10,
        ge=1,
        le=100,
        description="Connection pool size",
    )
    pool_timeout: int = Field(
        default=30,
        ge=1,
        description="Pool connection timeout in seconds",
    )
    echo: bool = Field(
        default=False,
        description="Echo SQL statements (for debugging)",
    )

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        """Validate database URL format."""
        valid_schemes = ("postgresql://", "mysql://", "sqlite://")
        if not any(v.startswith(scheme) for scheme in valid_schemes):
            raise ValueError(f"Database URL must start with one of: {valid_schemes}")
        return v


class RedisSettings(BaseSettings):
    """Redis configuration."""
    model_config = SettingsConfigDict(env_prefix="REDIS_")

    url: str = Field(
        default="redis://localhost:6379/0",
        description="Redis connection URL",
    )
    max_connections: int = Field(
        default=10,
        ge=1,
        description="Maximum connections in pool",
    )


class ApiSettings(BaseSettings):
    """External API configuration."""
    model_config = SettingsConfigDict(env_prefix="API_")

    base_url: str = Field(
        ...,
        description="External API base URL",
    )
    timeout: int = Field(
        default=30,
        ge=1,
        description="Request timeout in seconds",
    )
    api_key: str = Field(
        default="",
        description="API key for authentication",
    )
    retry_attempts: int = Field(
        default=3,
        ge=0,
        le=10,
        description="Number of retry attempts",
    )


class FeatureFlags(BaseSettings):
    """Feature flag configuration."""
    model_config = SettingsConfigDict(env_prefix="FEATURE_")

    new_dashboard: bool = Field(
        default=False,
        description="Enable new dashboard UI",
    )
    analytics: bool = Field(
        default=True,
        description="Enable analytics tracking",
    )
    beta_features: bool = Field(
        default=False,
        description="Enable beta features",
    )


class AppSettings(BaseSettings):
    """
    Main application configuration.

    Loads from environment variables with optional .env file support.
    """
    model_config = SettingsConfigDict(
        env_file=".env",
        env_file_encoding="utf-8",
        case_sensitive=False,
        extra="ignore",
    )

    # Application
    app_name: str = Field(
        default="MyApp",
        description="Application name",
    )
    env: Environment = Field(
        default=Environment.DEVELOPMENT,
        alias="NODE_ENV",
        description="Application environment",
    )
    debug: bool = Field(
        default=False,
        description="Enable debug mode",
    )
    log_level: LogLevel = Field(
        default=LogLevel.INFO,
        description="Logging level",
    )

    # Server
    host: str = Field(
        default="0.0.0.0",
        description="Server bind host",
    )
    port: int = Field(
        default=8000,
        ge=1,
        le=65535,
        description="Server bind port",
    )

    # Security
    secret_key: str = Field(
        ...,
        min_length=32,
        description="Secret key for encryption/signing",
    )
    allowed_hosts: list[str] = Field(
        default=["localhost", "127.0.0.1"],
        description="Allowed host headers",
    )
    cors_origins: list[str] = Field(
        default=["http://localhost:3000"],
        description="Allowed CORS origins",
    )

    # Nested configurations
    database: DatabaseSettings = Field(default_factory=DatabaseSettings)
    redis: RedisSettings = Field(default_factory=RedisSettings)
    api: ApiSettings = Field(default_factory=ApiSettings)
    features: FeatureFlags = Field(default_factory=FeatureFlags)

    @model_validator(mode="after")
    def validate_production_settings(self) -> "AppSettings":
        """Validate settings for production environment."""
        if self.env == Environment.PRODUCTION:
            if self.debug:
                raise ValueError("Debug mode must be disabled in production")
            if len(self.secret_key) < 64:
                raise ValueError("Secret key must be at least 64 chars in production")
            if "*" in self.allowed_hosts:
                raise ValueError("Wildcard host not allowed in production")
        return self

    @property
    def is_development(self) -> bool:
        """Check if running in development mode."""
        return self.env == Environment.DEVELOPMENT

    @property
    def is_production(self) -> bool:
        """Check if running in production mode."""
        return self.env == Environment.PRODUCTION

    @property
    def is_testing(self) -> bool:
        """Check if running in test mode."""
        return self.env == Environment.TEST


@lru_cache
def get_settings() -> AppSettings:
    """
    Get cached application settings.

    Uses lru_cache to ensure settings are only loaded once.
    """
    return AppSettings()


def validate_settings() -> None:
    """
    Validate settings at startup.

    Call this early in application initialization to fail fast
    on configuration errors.
    """
    try:
        settings = get_settings()
        print(f"Configuration loaded: env={settings.env.value}, debug={settings.debug}")
    except Exception as e:
        print(f"Configuration validation failed: {e}")
        raise SystemExit(1) from e


# CUSTOMIZE: Example usage
# from config_loader import get_settings, validate_settings
#
# # At application startup
# validate_settings()
#
# # Anywhere in your application
# settings = get_settings()
# print(f"Running on {settings.host}:{settings.port}")
# print(f"Database: {settings.database.url}")
# print(f"Feature new_dashboard: {settings.features.new_dashboard}")

File: snippets/typescript/config-loader.ts
/**
 * USE: When loading and validating configuration from environment variables
 * REQUIRES: TypeScript 4.5+, Node.js (process.env)
 * PATTERN: Configuration, Validation
 */

// Configuration validation error
class ConfigError extends Error {
  constructor(
    public readonly key: string,
    message: string,
    public readonly value?: unknown
  ) {
    super(`Configuration error for "${key}": ${message}`);
    this.name = 'ConfigError';
  }
}

// Environment variable reader with type coercion
type EnvReader<T> = {
  required: () => T;
  optional: (defaultValue: T) => T;
  optionalNullable: () => T | undefined;
};

function getString(key: string): EnvReader<string> {
  const value = process.env[key];

  return {
    required: () => {
      if (value === undefined || value === '') {
        throw new ConfigError(key, 'Required environment variable is not set');
      }
      return value;
    },
    optional: (defaultValue: string) => value ?? defaultValue,
    optionalNullable: () => value,
  };
}

function getNumber(key: string): EnvReader<number> {
  const rawValue = process.env[key];

  const parse = (): number | undefined => {
    if (rawValue === undefined || rawValue === '') return undefined;
    const num = Number(rawValue);
    if (Number.isNaN(num)) {
      throw new ConfigError(key, `Expected a number, got "${rawValue}"`);
    }
    return num;
  };

  return {
    required: () => {
      const value = parse();
      if (value === undefined) {
        throw new ConfigError(key, 'Required environment variable is not set');
      }
      return value;
    },
    optional: (defaultValue: number) => parse() ?? defaultValue,
    optionalNullable: () => parse(),
  };
}

function getBoolean(key: string): EnvReader<boolean> {
  const rawValue = process.env[key];

  const parse = (): boolean | undefined => {
    if (rawValue === undefined || rawValue === '') return undefined;
    const lower = rawValue.toLowerCase();
    if (['true', '1', 'yes', 'on'].includes(lower)) return true;
    if (['false', '0', 'no', 'off'].includes(lower)) return false;
    throw new ConfigError(key, `Expected a boolean, got "${rawValue}"`);
  };

  return {
    required: () => {
      const value = parse();
      if (value === undefined) {
        throw new ConfigError(key, 'Required environment variable is not set');
      }
      return value;
    },
    optional: (defaultValue: boolean) => parse() ?? defaultValue,
    optionalNullable: () => parse(),
  };
}

function getEnum<T extends string>(key: string, allowedValues: readonly T[]): EnvReader<T> {
  const rawValue = process.env[key];

  const parse = (): T | undefined => {
    if (rawValue === undefined || rawValue === '') return undefined;
    if (!allowedValues.includes(rawValue as T)) {
      throw new ConfigError(
        key,
        `Expected one of [${allowedValues.join(', ')}], got "${rawValue}"`
      );
    }
    return rawValue as T;
  };

  return {
    required: () => {
      const value = parse();
      if (value === undefined) {
        throw new ConfigError(key, 'Required environment variable is not set');
      }
      return value;
    },
    optional: (defaultValue: T) => parse() ?? defaultValue,
    optionalNullable: () => parse(),
  };
}

function getUrl(key: string): EnvReader<string> {
  const rawValue = process.env[key];

  const parse = (): string | undefined => {
    if (rawValue === undefined || rawValue === '') return undefined;
    try {
      new URL(rawValue);
      return rawValue;
    } catch {
      throw new ConfigError(key, `Expected a valid URL, got "${rawValue}"`);
    }
  };

  return {
    required: () => {
      const value = parse();
      if (value === undefined) {
        throw new ConfigError(key, 'Required environment variable is not set');
      }
      return value;
    },
    optional: (defaultValue: string) => parse() ?? defaultValue,
    optionalNullable: () => parse(),
  };
}

// CUSTOMIZE: Define your application configuration
const Environment = ['development', 'staging', 'production', 'test'] as const;
type Environment = typeof Environment[number];

interface AppConfig {
  // Application
  env: Environment;
  port: number;
  host: string;
  logLevel: 'debug' | 'info' | 'warn' | 'error';

  // Database
  databaseUrl: string;
  databasePoolSize: number;

  // External services
  apiBaseUrl: string;
  apiTimeout: number;

  // Feature flags
  featureNewDashboard: boolean;
  featureAnalytics: boolean;
}

// CUSTOMIZE: Load your configuration
function loadConfig(): AppConfig {
  return {
    // Application
    env: getEnum('NODE_ENV', Environment).optional('development'),
    port: getNumber('PORT').optional(3000),
    host: getString('HOST').optional('0.0.0.0'),
    logLevel: getEnum('LOG_LEVEL', ['debug', 'info', 'warn', 'error'] as const).optional('info'),

    // Database
    databaseUrl: getString('DATABASE_URL').required(),
    databasePoolSize: getNumber('DATABASE_POOL_SIZE').optional(10),

    // External services
    apiBaseUrl: getUrl('API_BASE_URL').required(),
    apiTimeout: getNumber('API_TIMEOUT_MS').optional(30000),

    // Feature flags
    featureNewDashboard: getBoolean('FEATURE_NEW_DASHBOARD').optional(false),
    featureAnalytics: getBoolean('FEATURE_ANALYTICS').optional(true),
  };
}

// Singleton config instance
let configInstance: AppConfig | null = null;

function getConfig(): AppConfig {
  if (!configInstance) {
    configInstance = loadConfig();
  }
  return configInstance;
}

// Reset config (useful for testing)
function resetConfig(): void {
  configInstance = null;
}

// Validate all required config at startup
function validateConfig(): void {
  try {
    loadConfig();
    console.log('Configuration validated successfully');
  } catch (error) {
    if (error instanceof ConfigError) {
      console.error(`Configuration validation failed: ${error.message}`);
      process.exit(1);
    }
    throw error;
  }
}

export {
  loadConfig,
  getConfig,
  resetConfig,
  validateConfig,
  getString,
  getNumber,
  getBoolean,
  getEnum,
  getUrl,
  ConfigError,
  AppConfig,
  Environment,
};
